### PR TITLE
WIP - Cling unlock execution

### PR DESCRIPTION
**Pull request by Axel-Naumann**
Can one of the threading people cross-check that this does what it's supposed to do? I don't have a good, simple test case.

### CODE REVIEW

**Comment by dpiparo** [comment]
Hi @Axel-Naumann ,
a trivial use case could be:
```
root  -e "ROOT::EnableThreadSafety();std::thread([](){gInterpreter->ProcessLine(\"printf(\\\"Stuck\\\")\");})" -b -q
```

**Comment by Axel-Naumann** [comment]
@dpiparo you forgot to `join` - for both with and without my patch I get
```
root  -e "ROOT::EnableThreadSafety();std::thread([](){gInterpreter->ProcessLine(\"printf(\\\"Stuck\\\")\");}).join()" -b -q
Stuck(int) 5
```
So I still don't have a good way of determining whether this patch fixes the issue.

**Comment by Axel-Naumann** [comment]
I tried
```
root.exe -l -b -q -e "ROOT::EnableThreadSafety();auto w = [&](){gInterpreter->ProcessLine(\"gInterpreterMutex->Lock();gInterpreterMutex->UnLock();\");};std::thread t(w);t.join()"
```
and that works both with and without this PR. (Without the `UnLock()` I deadlock coming back from userland with this PR.) I.e. I don't have a way yet to verify this patch.

**Comment by dpiparo** [comment]
Hi,

indeed it does not work if it is one single line.
Try it as a macro:
```
void a(){

ROOT::EnableThreadSafety();

auto w = [&](){gInterpreter->ProcessLine("gInterpreterMutex->Lock();");};
std::thread t(w);
t.join();

}
```

**Comment by Axel-Naumann** [comment]
This code
```
#include "TROOT.h"
#include "TInterpreter.h"
#include <thread>
void clinglock() {
  ROOT::EnableThreadSafety();
  auto w = [&](){gInterpreter->ProcessLine("gInterpreterMutex->Lock();gInterpreterMutex->UnLock();");};
  std::thread t(w);
  t.join();
}
```
works both before and after my patch, both compiled and interpreted. I.e. I still don't know what I'm fixing :-(

**Comment by dpiparo** [comment]
Hi,

this does trigger a deadlock on my machine. I think it's best to look at this together...

**Comment by pcanal** [comment]
> and that works both with and without this PR.

Yes that it is expected.  It is fundamental that the enable thread safety and the thread lauching are in *two* different Cling transaction.   This is because before the thread safely in enabled there is (of course) no lock being taken.
So try.
```
root.exe -b -l -e  "ROOT::EnableThreadSafety();" -e "std::thread([](){gInterpreter->ProcessLine(\"printf(\\\"Stuck\\\")\");}).join()" -q
```
Cheers,
Philippe.

**Comment by pcanal** [comment]
Also the file below (deadlock.cxx) is a Cling standalone reproducer of the mechanics of the problem.
To use:

```
root [0] .L deadlock.cxx+
root [1] deadlock(false); // Do not emulate the 'protection' the internal of cling
WWoouulldd  eexxeeccuuttee::  WWoorrkk((**((ssttdd::::ssttrriinngg**))00xx770000000000100830ee3300));;

Work being done within the lock of Second
Work being done within the lock of First
We successfully reached the end.
root [2] deadlock(true); // Emulate the 'protection' the internal of cling
WWoouulldd  eexxeeccuuttee::  WWoorrkk((**((ssttdd::::ssttrriinngg**))00xx770000000000100830ee3300));;

..... Process is now deadlocked .....


i.e. the point is that the same lock taken by the user code (Function Work) must be the same as the one used to 'protect the internal of cling'
```




```
#include <mutex>
#include <thread>
#include <iostream>
#include <sstream>

std::mutex gMutex;


void Work(const std::string &name)
{
   // simulate a long page fetch
   std::this_thread::sleep_for(std::chrono::seconds(2));
   std::string result = "fake content";

   // This lock would be taken indirectly by any access to the global database
   // (TClass, TCling, etc..)
   std::lock_guard<std::mutex> guard(gMutex);
   cerr << "Work being done within the lock of " << name << '\n';
}

void IndirectWork(const std::string &name, bool takelock)
{

   std::stringstream cmd;
   cmd << "Work(*(std::string*)";
   cmd << (void*)&name;
   cmd <<  ");" ;
   cerr << "Would execute: " << cmd.str()<< '\n';

   if (takelock) {
      // This lock is taken by gROOT->ProcessLine.
      std::lock_guard<std::mutex> guard(gMutex);
      /*
       interpreter->execute(cmd.str().c_str());
       */
      Work(name);
   } else {
      /*
       interpreter->execute(cmd.str().c_str());
       */
      Work(name);
   }
}

void deadlock(bool takelock = true)
{

   std::thread t1(IndirectWork, "First", takelock);
   std::thread t2(IndirectWork, "Second", takelock);

   t1.join();
   t2.join();
   cerr << "We successfully reached the end.\n";
}
```

File: core/metacling/src/TClingCallbacks.cxx [hunk 1]
```diff
@@ -783,3 +785,11 @@ void TClingCallbacks::LibraryUnloaded(const void* dyLibHandle,
 void TClingCallbacks::PrintStackTrace() {
    TCling__PrintStackTrace();
 }
+
+void TClingCallbacks::EnteringUserCode() {
+  TCling__UnlockInterpreterMutex();
```
Comment (pcanal) [inline core/metacling/src/TClingCallbacks.cxx]: Add
```
// We can safely assume that if the lock exist already when we are in Cling code,
// then the lock has (or should been taken) already.  So it is fair to unlock it.
```
However if the mutex is still null here we should record that information has it means we ought not to unlock it on exit ....

File: core/metacling/src/TClingCallbacks.cxx [hunk 2]
```diff
@@ -783,3 +785,11 @@ void TClingCallbacks::LibraryUnloaded(const void* dyLibHandle,
...
+}
+
+void TClingCallbacks::ReturnedFromUserCode() {
+  TCling__LockInterpreterMutex();
```
Comment (pcanal) [inline core/metacling/src/TClingCallbacks.cxx]: The unlocking ought to be conditional upon the existence of the mutex when we entered and/or (which should be equivalent) on whether we unlocked it.

File: core/metacling/src/TCling.cxx [hunk 3]
```diff
@@ -351,6 +351,29 @@ void TCling__PrintStackTrace() {
 }
 
 ////////////////////////////////////////////////////////////////////////////////
+/// Lock the interpreter.
+
+extern "C"
+void TCling__RestoreInterpreterMutex(void* state) {
+   if (gInterpreterMutex && state) {
+      std::unique_ptr<TVirtualMutex::State> uniqueP{static_cast<TVirtualMutex::State*>(state)};
```
Comment (pcanal) [inline core/metacling/src/TCling.cxx]: Would it be more readable with
```
auto uniqueP{static_cast<TVirtualMutex::State*>(state)};
```

File: interpreter/cling/lib/Interpreter/IncrementalExecutor.h [hunk 4]
```diff
@@ -150,7 +156,7 @@ namespace cling {
     void setExternalIncrementalExecutor(IncrementalExecutor *extIncrExec) {
       m_externalIncrementalExecutor = extIncrExec;
     }
-
+    void setCallbacks(InterpreterCallbacks* callbacks) { m_Callbacks = callbacks; }
```
Comment (pcanal) [inline interpreter/cling/lib/Interpreter/IncrementalExecutor.h]: Now that it is settable, don't we have to deal elegantly with the fact that m_Callbacks might not be nullptr when setCallbacks is callled (i.e. should the be
```
  auto tmp = m_Callbacks;
  m_Callbacks = callbacks;
  return m_Callbacks;
```
or should we have a getCallbacks 

and/or something else?

**Comment by pcanal** [comment]
Did you try a pertubation of the example that make sure the nested-lock taking is induced by interpreter level static initialization?

File: core/metacling/src/TCling.cxx (hunk 3 above)
Comment (Axel-Naumann) [reply core/metacling/src/TCling.cxx]: More readable yes, but sadly also ill-formed: I need a `unique_ptr`.

Comment (pcanal) [reply core/metacling/src/TCling.cxx]: which proves it is hard to read .. I obviously did not really noticed the missing unique_ptr in the cast :( 
What about
```c++
void TCling__RestoreInterpreterMutex(void *state) {
   if (gInterpreterMutex && state) {
      auto typedState = static_cast<TVirtualMutex::State*>(state);
      std::unique_ptr<TVirtualMutex::State> uniqueP{typeState};
      gInterpreterMutex->Restore(std::move(uniqueP));
   }
}
```

and you shouldn't use that but the following ought to work also

```
void TCling__RestoreInterpreterMutex(void* state) {
   if (gInterpreterMutex && state) {
      // unique_ptr is guaranteed to be the same size and content as void*
      auto uniquePtrPtr = reinterpret_cast<std::unique_ptr<TVirtualMutex::State> >( &state );
      gInterpreterMutex->Restore(std::move(*uniquePtrPtr));
   }
}
```

File: interpreter/cling/lib/Interpreter/IncrementalExecutor.h (hunk 4 above)
Comment (Axel-Naumann) [reply interpreter/cling/lib/Interpreter/IncrementalExecutor.h]: Let's not overdesign this - it's an internal interface, and we can add it when we need it. Right now it gets called exactly once.

**Comment by Axel-Naumann** [comment]
@phsft-bot build!

**Comment by Axel-Naumann** [comment]
Hand-merged (rebase-with-merge-commit).

**Pull request closed without merging**